### PR TITLE
Add `LimitWriteBucket`

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Adds a `LimitWriteBucket` that limits the total size of bytes that can be written to the bucket.

### CODE REVIEW

File: private/pkg/storage/limit.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	return newLimitedWriter(writeBucket, limit)
+}
+
+type limitedWriteBucket struct {
+	WriteBucket
+	currentSize *atomic.Int64
+	limit       int64
+}
+
+func newLimitedWriter(bucket WriteBucket, limit int64) *limitedWriteBucket {
```
Comment (bufdev) [inline private/pkg/storage/limit.go:38]: `newLimitWriteBucket

Comment (bufdev) [inline private/pkg/storage/limit.go:38]: I would probably just make this an `int`, sort of a standard we have with interface types, and then correct <0 to be 0, ie `if limit < 0 { limit = 0 }`

Comment (srikrsna-buf) [reply private/pkg/storage/limit.go]: Updated the public function to accept an `int`

File: private/pkg/storage/limit.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		WriteObjectCloser: writeObjectCloser,
+		bucketSize:        bucketSize,
+		limit:             limit,
+	}
+}
+
+func (o *limitedWriteObjectCloser) Write(p []byte) (n int, err error) {
+	newSize := o.bucketSize.Add(int64(len(p)))
+	if newSize > o.limit {
+		return 0, fmt.Errorf("limit writer: write limit reached: limit: %d, exceeded by: %d", o.limit, newSize-o.limit)
```
[the remark this answers is not in the document]
Comment (srikrsna-buf) [reply private/pkg/storage/limit.go:81]: Yeah, I was going to update it with a variant of the second approach you suggested.

File: private/pkg/storage/limit.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	newBucketSize := o.bucketSize.Add(writeSize)
+	if newBucketSize > o.limit {
+		o.bucketSize.Sub(writeSize)
+		return 0, &errWriteLimitReached{
+			Limit:       o.limit,
+			ExceedingBy: newBucketSize - o.limit,
+		}
+	}
+	writtenSize, err := o.WriteObjectCloser.Write(p)
+	if writtenSize < int(writeSize) {
```
Comment (twilly) [inline private/pkg/storage/limit.go:88]: Casting int -> int64 -> int is a bit confusing. If we're going to use int64 for counting over plain int, then it's better to always cast upward (int -> int64).

```go
if int64(writtenSize) < writeSize {
```

File: private/pkg/storage/limit.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	limit int64,
+) *limitedWriteObjectCloser {
+	return &limitedWriteObjectCloser{
+		WriteObjectCloser: writeObjectCloser,
+		bucketSize:        bucketSize,
+		limit:             limit,
+	}
+}
+
+func (o *limitedWriteObjectCloser) Write(p []byte) (n int, err error) {
```
Comment (twilly) [inline private/pkg/storage/limit.go:77]: No need for named return parameters if you don't use the names. :)

```suggestion
func (o *limitedWriteObjectCloser) Write(p []byte) (int, error) {
```

**Review by twilly — COMMENTED** [review]
Looking pretty good. A few tests and this thing should be good to go.

**Review by doriable — APPROVED** [review]

**Pull request merged**